keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md

```diff
@@ -0,0 +1,297 @@
+    * API group name: `{component}.config.k8s.io`
+    * Kind name: `{Component}Configuration`
+    * Code location:
+        * External types: `k8s.io/{component}/config/{version}/types.go`
+            * Like `k8s.io/api`
+        * Internal types: `k8s.io/kubernetes/pkg/{component}/apis/config`
+            * Alternatives, if applicable
+                * `k8s.io/{component}/pkg/apis/config` (preferred)
+                * `k8s.io/kubernetes/cmd/{component}/app/apis/config`
+            * If dependencies allow it, we can move them to `k8s.io/{component}/pkg/apis/config/types.go`. Not having the external types there is intentional because the `pkg/` package tree is considered as "on-your-own-risks / no code compatibility guarantees", while `config/` is considered as a code API.
+        * Internal scheme package: `k8s.io/kubernetes/pkg/{component}/apis/config/scheme/scheme.go`
+            * The scheme package should expose `Scheme *runtime.Scheme`, `Codecs *serializer.CodecFactory`, and `AddToScheme(*runtime.Scheme)`, and have an `init()` method that runs `AddToScheme(Scheme)`
+    * For the move to a staging repo to be possible, the external API package must not depend on the core repo.
+        * Hence, all non-staging repo dependencies need to be removed/resolved before the package move.
+    * Conversions from the external type to the internal type will be kept in `{internal_api_path}/{external_version}`, like for `k8s.io/api`
+        * Defaulting code will be kept in this package, next to the conversions
+* Create a "shared types"-package with structs generic to all or many componentconfig API groups, in the `k8s.io/apimachinery`, `k8s.io/apiserver` and `k8s.io/controller-manager` repos, depending on the struct.
+    * Location: `k8s.io/{apimachinery,apiserver,controller-manager}/pkg/apis/config/{,v1alpha1}`
+    * These aren’t "real" API groups, but have both internal and external versions
+    * Defaulting, conversion and internal types are published to the staging repo.
+* Remove the monolithic `componentconfig/v1alpha1` API group
+* Enable the staging bot to create the Github repos
+
+### Migration strategy per component or k8s.io repo
+
+#### k8s.io/apimachinery changes
+
+* **Not a "real" API group, instead shared packages only with both external and internal types.**
+* **External Package with defaulting & conversions**: `k8s.io/apimachinery/pkg/apis/config/v1alpha1/types.go`
+* **Internal Package**: `k8s.io/apimachinery/pkg/apis/config/types.go`
+* Structs to be hosted:
+    * ClientConnectionConfiguration
+* Assignee: @hanxiaoshuai
+
+#### k8s.io/apiserver changes
+
+* **Not a "real" API group, instead shared packages only with both external and internal types.**
+* **External Package with defaulting & conversions**: `k8s.io/apiserver/pkg/apis/config/v1alpha1/types.go`
+* **Internal Package**: `k8s.io/apiserver/pkg/apis/config/types.go`
+* Structs to be hosted:
+    * LeaderElectionConfiguration
```

Comment:
this would seem to fit under apimachinery, not apiserver (looking at the existing consumers)

Reply:
For the lack of a better place right now, I leave this in apiserver. We can move that elsewhere at another point in time if we find a better place. Apimachinery and Apiserver are equally wrong.

Reply:
And the other place, which would be `k8s.io/controller-manager` is also kinda wrong as any component can do leader election, not only custom controller-managers.